# Assignment 2

## Instructions - Read this first!

This is an individual homework assignment. This means that:

- You may discuss the problems in this assignment with other students in this course and your instructor/TA, but YOUR WORK MUST BE YOUR OWN.
- Do not show other students code or your own work on this assignment.
- You may consult external references, but not actively receive help from individuals not involved in this course.
- Cite all references outside of the course you used, including conversations with other students which were helpful. (This helps us give credit where it is due!). All references must use a commonly accepted reference format, for example, APA or IEEE (or another citation style of your choice).

If any of these rules seem ambiguous, please check with with your instructor for help interpreting them.

We suggest completing this assignment using the provided notebook. Each question should be answered using a SQL query (or combination or SQL queries) unless the text indicates that you may do something else. You may submit your queries embedded in Python, using SQLAlchemy or the MySQL Connector, or as plain text in Markdown.

## When you submit your work

Your submission will be graded manually. To ensure that everything goes smoothly, please follow these instructions to prepare your notebook for submission to the D2L Dropbox for Assignment 2:

- Please remove any print statments used to test your work (you can comment them out)
- Please provide your solutions where asked; please do not alter any other parts of this notebook.
- If you need to add cells to test your code please move them to the end of the notebook before submission- or you may included your commented out answers and tests in the cells provided

## Introduction

 In this assignment, we will focus familiarizing you with using SQL for data exploration, and continuing to cultivate a sense of curiosity about the datasets you encounter. We will be using a table generated by Statistics Canada (2022), based upon census data, about the knowledge of languages by area in Canada. This table has been pre-cleaned (although potentially not entirely) so that you can work on the actual assignment more quickly.
 
 To begin, start by importing the provided CSV into your own SQL database using SQLAlchemy, by filling in the lines below:

In [1]:
print('hello')
print(12)

hello
12


In [1]:
import pandas as pd
import sqlalchemy as sq
import mysql.connector
from pandas.io import sql

# read in your CSV as a dataframe
a2df = pd.read_csv("language_speakers.csv")
a2df.head()
# connect to your database; include a cell at the bottom of this notebook to dispose of your engine object
engine = sq.create_engine('mysql+mysqlconnector://zheyu_song:5QFH73NUI@datasciencedb.ucalgary.ca/zheyu_song')
# write your dataframe into a table
a2df.to_sql('languages', engine )
# demonstrate that your import has been successful by reading your database table as a dataframe, 
languagesdf = pd.read_sql_table("languages", engine)
display(languagesdf.head())
# and print some information (not the entire table) about your second dataframe
display(languagesdf.info())



index  REF_DATE     GEO           DGUID Statistics (3)  \
0      0      2021  Canada  2021A000011124          Count   
1      1      2021  Canada  2021A000011124          Count   
2      2      2021  Canada  2021A000011124          Count   
3      3      2021  Canada  2021A000011124          Count   
4      4      2021  Canada  2021A000011124          Count   

     Knowledge of languages (324)  \
0  Total - Knowledge of languages   
1  Total - Knowledge of languages   
2  Total - Knowledge of languages   
3              Official languages   
4              Official languages   

  Single and multiple responses of knowledge of languages (3) SCALAR_FACTOR  \
0  Total - Single and multiple responses of knowl...                  units   
1         Single responses of knowledge of languages                  units   
2       Multiple responses of knowledge of languages                  units   
3  Total - Single and multiple responses of knowl...                  units   
4         Single responses of knowledge of languages                  units   

   SCALAR_ID   COORDINATE     VALUE  
0          0  1.1.1.1.1.1  36328480  
1          0  1.1.1.1.1.2  21377910  
2          0  1.1.1.1.1.3  14950565  
3          0  1.1.1.1.2.1  35644655  
4          0  1.1.1.1.2.2  20797495

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7536 entries, 0 to 7535
Data columns (total 11 columns):
 #   Column                                                       Non-Null Count  Dtype 
---  ------                                                       --------------  ----- 
 0   index                                                        7536 non-null   int64 
 1   REF_DATE                                                     7536 non-null   int64 
 2   GEO                                                          7536 non-null   object
 3   DGUID                                                        7536 non-null   object
 4   Statistics (3)                                               7536 non-null   object
 5   Knowledge of languages (324)                                 7536 non-null   object
 6   Single and multiple responses of knowledge of languages (3)  7536 non-null   object
 7   SCALAR_FACTOR                                                7536 non-null   object
 8 

None

## Part A: Warm-up Questions (12 marks)

Answer the questions below, including the queries you used where necessary. Not all questions will require writing a SQL query to answer.

**(1 mark)**

How many records are there in total?

In [17]:
query_table = pd.read_sql_query('SELECT COUNT(*) as records FROM languages;', engine)
print(query_table)


   records
0     7536


**(1 mark)**

How many different languages are spoken in Canada?

In [18]:
create_statement = '''select COUNT(DISTINCT `Knowledge of languages (324)`) as CountL from languages 
where  `Knowledge of languages (324)` not LIKE '%languages';'''

query1_table = pd.read_sql_query(create_statement, engine)
print(query1_table)


   CountL
0     263


**(1 mark)**

Is there data for all languages, for each province or territory?

In [19]:
query1_table = pd.read_sql_query('SELECT GEO AS PT FROM languages GROUP BY GEO;', engine)
print(query1_table)
#The table/dataset contains data from all ten province and three territory.
#But I can not say there is data for all languages due to the limitation of survery.
#Thus, there is not data for all languages, for each province or territory.

                           PT
0                     Alberta
1            British Columbia
2                      Canada
3                    Manitoba
4               New Brunswick
5   Newfoundland and Labrador
6       Northwest Territories
7                 Nova Scotia
8                     Nunavut
9                     Ontario
10       Prince Edward Island
11                     Quebec
12               Saskatchewan
13                      Yukon


**(8 marks)**

Explain what each of the following columns is used for. You may use the original page to guide your explanation (but you should cite it)
* GEO
* Statistics
* Knowledge of languages
* Single and multiple responses of knowledge of languages (2 marks)
* COORDINATE (2 marks)
* VALUE

* GEO: GEO contains data for all Canadian provinces and territories as well as national data, which allows us to study the problem geographically.
* Statistics: Statistics contains three Count methods, they are Total - Single and multiple responses of knowledge of languages, Single responses of knowledge of languages and Multiple responses of knowledge of languages.
* Knowledge of languages: It contains different kinds of set of languages with hierarchy of categories such as Official languages and non Official languages, under each category contains sub categories of languages.
* Single and multiple responses of knowledge of languages:
Single responses of knowledge of languages: Number of people speaks only speak the language. A single language response occurs when a respondent provides one language only (Statistics Canada,2022).
Multiple responses of knowledge of languages: Number of people speaks the language while speaks some other languages. A multiple language response occurs when a respondent provides two or more languages (Statistics Canada,2022).
Total - Single and multiple responses of knowledge of languages: Number of people speaks this language. No strings attached.
* COORDINATE: A numerical representation of the dataset. The first item has 14 unique number represents the column GEO which contains 13  Canadian provinces and territories and Canada countrariwise. The penultimate item represents the colume Knowledge of languages. The last item represents the colume Single responses of knowledge of languages. 
* VALUE: The value of the each sub category of Single and multiple responses of knowledge of languages colume.

**(1 mark)**

Is it possible to tell how many individuals live in Canada based on this dataset? Why or why not?

It is not possible to tell how many individuals live in Canada based on this dataset. Usually a census is required for demographics. From the data the sample size of 36,328,480 people in canada was selected. Which is not a demographics.

## Part B: Simple questions (10 marks) 

For these queries, run a query which provides the answer.

**(2 marks)**

Which province or territory has the highest number of individuals who can speak multiple languages?

In [20]:
create_statement = '''select GEO, count(`Single and multiple responses of knowledge of languages (3)`) as Countml from languages
where `Single and multiple responses of knowledge of languages (3)` = 'Multiple responses of knowledge of languages'
group by GEO
ORDER BY COUNT(`Single and multiple responses of knowledge of languages (3)`) DESC;'''

create_statement1 = '''select GEO, `VALUE` as number_multiple_languages_spaker from languages
WHERE `Knowledge of languages (324)` = 'Total - Knowledge of languages' 
AND `Single and multiple responses of knowledge of languages (3)` ='Multiple responses of knowledge of languages'
ORDER BY `VALUE` DESC;'''

query1_table = pd.read_sql_query(create_statement1, engine)
print(query1_table)

#The output shows that Ontario has the highest number of individuals who can speak multiple languages.


                          GEO  number_multiple_languages_spaker
0                      Canada                          14950565
1                     Ontario                           5855325
2                      Quebec                           4564580
3            British Columbia                           1870730
4                     Alberta                           1354235
5                    Manitoba                            458075
6               New Brunswick                            296330
7                Saskatchewan                            245430
8                 Nova Scotia                            178055
9   Newfoundland and Labrador                             43800
10       Prince Edward Island                             32980
11                    Nunavut                             25715
12      Northwest Territories                             13510
13                      Yukon                             11815


**(2 marks)**

Which provinces or territories has the most diverse number of languages spoken?

In [21]:
create_statement = '''select GEO, COUNT(DISTINCT `Knowledge of languages (324)`) as CountL from languages 
where  `Knowledge of languages (324)` not LIKE '%languages'
group by GEO
ORDER BY CountL DESC;'''


query1_table = pd.read_sql_query(create_statement, engine)
print(query1_table)

#The output shows that British Columbia has the most diverse number of languages spoken.


                          GEO  CountL
0                      Canada     263
1            British Columbia     252
2                     Alberta     235
3                     Ontario     234
4                      Quebec     220
5                    Manitoba     203
6                Saskatchewan     197
7                 Nova Scotia     169
8               New Brunswick     156
9   Newfoundland and Labrador     124
10      Northwest Territories     102
11       Prince Edward Island     100
12                      Yukon      94
13                    Nunavut      62


**(2 marks)**

Which provinces or territories have the most diverse number of Indigenous languages spoken?

In [22]:
create_statement = '''SELECT GEO, COUNT(DISTINCT `Knowledge of languages (324)`) AS good
from languages
where COORDINATE REGEXP '[.]([6-9]|[1-8][0-9]|9[0-6])[.][1-3]$'
and  `Knowledge of languages (324)` not LIKE '%languages'
group by GEO
ORDER BY good DESC;'''

query1_table = pd.read_sql_query(create_statement, engine)
print(query1_table)
display(query1_table)
#The output shows that British Columbia has the most diverse number of Indigenous languages spoken.

#https://www.regular-expressions.info/numericranges.html
#https://www.regextutorial.org/regex-for-numbers-and-ranges.php
#https://dev.mysql.com/doc/refman/8.0/en/regexp.html#operator_regexp

                          GEO  good
0                      Canada    73
1            British Columbia    63
2                     Alberta    47
3                     Ontario    44
4                      Quebec    33
5                    Manitoba    21
6                Saskatchewan    21
7       Northwest Territories    19
8                       Yukon    17
9                 Nova Scotia    13
10              New Brunswick    11
11  Newfoundland and Labrador     9
12                    Nunavut     7
13       Prince Edward Island     2


GEO  good
0                      Canada    73
1            British Columbia    63
2                     Alberta    47
3                     Ontario    44
4                      Quebec    33
5                    Manitoba    21
6                Saskatchewan    21
7       Northwest Territories    19
8                       Yukon    17
9                 Nova Scotia    13
10              New Brunswick    11
11  Newfoundland and Labrador     9
12                    Nunavut     7
13       Prince Edward Island     2

**(2 marks)**

As a percentage of all language speakers in each province or territory, how commonly is French spoken?

In [10]:
create_statement1 = '''
CREATE TEMPORARY TABLE IF NOT EXISTS table1 AS (select GEO, `VALUE` from languages
WHERE `Knowledge of languages (324)` = 'Total - Knowledge of languages' 
AND `Single and multiple responses of knowledge of languages (3)` = 'Total - Single and multiple responses of knowledge of languages'
ORDER BY `VALUE` DESC);
'''

create_statement2 = '''
CREATE TEMPORARY TABLE IF NOT EXISTS table2 AS (select GEO, `VALUE` from languages
WHERE `Knowledge of languages (324)` = 'French' 
AND `Single and multiple responses of knowledge of languages (3)` ='Total - Single and multiple responses of knowledge of languages'
ORDER BY `VALUE` DESC);
'''

sql.execute(create_statement1,engine)
sql.execute(create_statement2,engine)
query1_table = pd.read_sql_query("SELECT * FROM table1", engine)
print(query1_table)
query2_table = pd.read_sql_query("SELECT * FROM table2", engine)
#print(query2_table)

create_statement3 = '''
select t1.GEO, (t2.`VALUE` / t1.`VALUE` * 100) as Percentagefrenchspeker
FROM table1 t1 INNER JOIN table2 t2
ON t1.GEO = t2.GEO 
ORDER BY Percentagefrenchspeker DESC;

'''

query3_table = pd.read_sql_query(create_statement3, engine)
print(query3_table)

#https://stackoverflow.com/questions/22082215/dividing-two-fields-from-two-different-tables
#https://stackoverflow.com/questions/30945513/how-to-divide-two-values-in-same-column
#help from leo hao su

                          GEO     VALUE
0                      Canada  36328480
1                     Ontario  14031750
2                      Quebec   8308480
3            British Columbia   4915945
4                     Alberta   4177715
5                    Manitoba   1307190
6                Saskatchewan   1103205
7                 Nova Scotia    955855
8               New Brunswick    759195
9   Newfoundland and Labrador    502100
10       Prince Edward Island    150480
11      Northwest Territories     40380
12                      Yukon     39590
13                    Nunavut     36605
                          GEO  Percentagefrenchspeker
0                      Quebec                   93.72
1               New Brunswick                   41.86
2                      Canada                   29.08
3                       Yukon                   14.50
4        Prince Edward Island                   12.92
5                     Ontario                   11.05
6       Northwest Terr

**(2 marks)**

As a percentage of all Chinese speakers in the country, how many Chinese speakers do not speak Mandarin?

In [5]:
create_statement2 = '''
CREATE TEMPORARY TABLE IF NOT EXISTS chinese AS (select GEO, `VALUE` as AllChinese 
from languages
WHERE `Knowledge of languages (324)` = 'Chinese languages' 
AND `Single and multiple responses of knowledge of languages (3)` = 'Total - Single and multiple responses of knowledge of languages'
ORDER BY `VALUE` DESC);
'''
sql.execute(create_statement2,engine)
#query1_table = pd.read_sql_query("SELECT * FROM chinese", engine)
#print(query1_table)

create_statement1 = '''select t1.GEO, SUM(t1.`VALUE`) nonMandarin,t2.AllChinese AllChinese, (SUM(t1.`VALUE`) / t2.AllChinese) as PercentageNonMandarin
from languages t1 INNER JOIN chinese t2
ON t1.GEO = t2.GEO 
WHERE NOT `Knowledge of languages (324)` = 'Mandarin' 
AND `Single and multiple responses of knowledge of languages (3)` = 'Total - Single and multiple responses of knowledge of languages'
AND COORDINATE REGEXP '[.]29[2-9][.][1-3]$'
group by GEO
ORDER BY PercentageNonMandarin DESC;'''

query2_table = pd.read_sql_query(create_statement1, engine)
print(query2_table)

create_statement3 = '''
CREATE TEMPORARY TABLE IF NOT EXISTS cn AS (select SUM(`VALUE`) as AllChinese 
from languages
WHERE `Knowledge of languages (324)` = 'Chinese languages' 
AND `Single and multiple responses of knowledge of languages (3)` = 'Total - Single and multiple responses of knowledge of languages'
AND GEO = 'Canada'
);
'''
sql.execute(create_statement3,engine)
query3_table = pd.read_sql_query("SELECT * FROM cn", engine)
#print(query3_table)

create_statement4 = '''select t1.GEO, (SUM(t1.`VALUE`) / t2.AllChinese * 100) as 'PercentageNonMandarin(%)'
from languages t1 INNER JOIN chinese t2
ON t1.GEO = t2.GEO 
WHERE NOT `Knowledge of languages (324)` = 'Mandarin' 
AND `Single and multiple responses of knowledge of languages (3)` = 'Total - Single and multiple responses of knowledge of languages'
AND COORDINATE REGEXP '[.]29[2-9][.][1-3]$'
AND t1.GEO = 'Canada'
;'''

query4_table = pd.read_sql_query(create_statement4, engine)
print(query4_table)

                          GEO  nonMandarin  AllChinese  PercentageNonMandarin
0       Northwest Territories        195.0         310                 0.6290
1                       Yukon        300.0         480                 0.6250
2                     Alberta      83220.0      141465                 0.5883
3            British Columbia     281185.0      488470                 0.5756
4                     Nunavut         40.0          70                 0.5714
5                      Canada     819905.0     1528860                 0.5363
6                     Ontario     392810.0      734190                 0.5350
7                    Manitoba       9820.0       23210                 0.4231
8                      Quebec      42205.0      106915                 0.3948
9                Saskatchewan       5870.0       15620                 0.3758
10  Newfoundland and Labrador        505.0        1525                 0.3311
11                Nova Scotia       2310.0        9775          

## Part C: Detailed analysis (20 marks)

Now consider being given a task to make sense of the distribution of languages spoken in Canada using this dataset.

**Question 1 (4 marks)**

Create two guiding questions to use in your analysis, and include them below as Markdown. As a starting point (and remember you are not limited to only these!), you may want to consider the following ideas:
- focus on a specific set of languages which are interesting to you
- consider the effect of language speakers who speak only one or multiple languages
- differences between specific provinces and territories in terms of patterns of language speakers

1. Mandarin and Cantonese are the two most widely spoken Chinese languages in Canada. Which language is used more in each province or territory?
2. For Mandarin and Cantonese, in each province or territory which language has the highest Single responses of knowledge of languages rate ?

**Question 2 (12 marks)** 

Write at least four queries (two for each question) which you believe will address one of your guiding questions. Clearly indicate which queries address your questions. You may wish to include a comment to explain why this query will help address your question.

In [19]:
#question 1 

create_statement4 = '''select GEO, `Knowledge of languages (324)`, VALUE
FROM languages
WHERE `Single and multiple responses of knowledge of languages (3)` = 'Total - Single and multiple responses of knowledge of languages'
AND `Knowledge of languages (324)` = 'Mandarin'  OR  `Knowledge of languages (324)` = 'Yue (Cantonese)' 
AND `Single and multiple responses of knowledge of languages (3)` = 'Total - Single and multiple responses of knowledge of languages'
;'''
#this query give the the number of Mandarin and Cantonese speaker in each province and territory.

query4_table = pd.read_sql_query(create_statement4, engine)
print(query4_table)

create_statement = '''select GEO, `Knowledge of languages (324)`, VALUE,
RANK() OVER (partition by GEO order by VALUE desc) AS ranking
from languages 

WHERE COORDINATE REGEXP '[.]29[2-9][.]1$'
AND `Knowledge of languages (324)` = 'Mandarin'  OR  COORDINATE REGEXP '[.]29[2-9][.]1$' and `Knowledge of languages (324)` = 'Yue (Cantonese)' 

;'''
#this query give the rank of the number of chinese speaker in each province and territory.

query_table = pd.read_sql_query(create_statement, engine)
print(query_table)


#From the SQL output We can analyze that 
#In Yukon, the number of Cantonese speaker is larger than the number of Mandarin speaker.
#In Northwest Territories, The number of people speaking Cantonese and Mandarin is the same.
#In all other province and territory, the number of people speaking Mandarin is greater than the number of people speaking Cantonese.

                          GEO Knowledge of languages (324)   VALUE
0                      Canada                     Mandarin  987300
1                      Canada              Yue (Cantonese)  724925
2   Newfoundland and Labrador                     Mandarin    1165
3   Newfoundland and Labrador              Yue (Cantonese)     395
4        Prince Edward Island                     Mandarin    2945
5        Prince Edward Island              Yue (Cantonese)     445
6                 Nova Scotia                     Mandarin    8525
7                 Nova Scotia              Yue (Cantonese)    1805
8               New Brunswick                     Mandarin    3185
9               New Brunswick              Yue (Cantonese)     585
10                     Quebec                     Mandarin   80520
11                     Quebec              Yue (Cantonese)   34025
12                    Ontario                     Mandarin  467420
13                    Ontario              Yue (Cantonese)  35

In [34]:
#Question 2
create_statement1 = '''
CREATE TEMPORARY TABLE IF NOT EXISTS singlem AS (select GEO, `VALUE` from languages
WHERE `Knowledge of languages (324)` = 'Mandarin' 
AND `Single and multiple responses of knowledge of languages (3)` = 'Single responses of knowledge of languages'
ORDER BY `VALUE` DESC);
'''
#this query CREATE a table of number of Mandarin speaker who speaks only Mandarin in each province and territory.

create_statement2 = '''
CREATE TEMPORARY TABLE IF NOT EXISTS totalm AS (select GEO, `VALUE` from languages
WHERE `Knowledge of languages (324)` = 'Mandarin'
AND `Single and multiple responses of knowledge of languages (3)` ='Total - Single and multiple responses of knowledge of languages'
ORDER BY `VALUE` DESC);
'''
#this query CREATE a table of number of Mandarin speaker who speaks Mandarin and some other languages in each province and territory.


sql.execute(create_statement1,engine)
sql.execute(create_statement2,engine)
query1_table = pd.read_sql_query("SELECT * FROM singlem", engine)
#print(query1_table)
query2_table = pd.read_sql_query("SELECT * FROM totalm", engine)
#print(query2_table)

create_statement3 = '''
select t1.GEO, (t2.`VALUE` / t1.`VALUE`) as singlemandarinrate
FROM singlem t2 INNER JOIN totalm t1
ON t1.GEO = t2.GEO 
ORDER BY singlemandarinrate DESC;

'''
#this query calculated the rate of single Mandarin speaker


query3_table = pd.read_sql_query(create_statement3, engine)
print(query3_table)

create_statement1 = '''
CREATE TEMPORARY TABLE IF NOT EXISTS singlec AS (select GEO, `VALUE` from languages
WHERE `Knowledge of languages (324)` = 'Yue (Cantonese)'
AND `Single and multiple responses of knowledge of languages (3)` = 'Single responses of knowledge of languages'
ORDER BY `VALUE` DESC);
'''
#this query CREATE a table of number of Mandarin speaker who speaks only Cantonese in each province and territory.

create_statement2 = '''
CREATE TEMPORARY TABLE IF NOT EXISTS totalc AS (select GEO, `VALUE` from languages
WHERE `Knowledge of languages (324)` = 'Yue (Cantonese)'
AND `Single and multiple responses of knowledge of languages (3)` ='Total - Single and multiple responses of knowledge of languages'
ORDER BY `VALUE` DESC);
'''
#this query CREATE a table of number of Mandarin speaker who speaks Cantonese and some other languages in each province and territory.

sql.execute(create_statement1,engine)
sql.execute(create_statement2,engine)


create_statement3 = '''
select t1.GEO, (t2.`VALUE` / t1.`VALUE`) as singleCantoneserate
FROM singlec t2 INNER JOIN totalc t1
ON t1.GEO = t2.GEO 
ORDER BY singleCantoneserate DESC;

'''
#this query calculated the rate of single Cantonese speaker

query3_table = pd.read_sql_query(create_statement3, engine)
print(query3_table)

create_statement3 = '''
select t1.GEO, (t2.`VALUE` / t1.`VALUE`) as singleCantoneserate, (t4.`VALUE` / t3.`VALUE`) as singleMandarinrate, ((t4.`VALUE` / t3.`VALUE`) - (t2.`VALUE` / t1.`VALUE`)) as diff
FROM singlec t2 INNER JOIN totalc t1 ON t1.GEO = t2.GEO 
INNER JOIN singlem t4 ON t1.GEO = t4.GEO 
INNER JOIN totalm t3 ON t1.GEO = t3.GEO  
;

'''

#this query calculated the differemces of rate of single Cantonese speaker and rate of single Mandarin speaker


query3_table = pd.read_sql_query(create_statement3, engine)
print(query3_table)



#From the SQL output We can analyze that 
#In Ontario,Newfoundland and Labrador,Saskatchewan and New Brunswick, the rate of single Mandarin speaker is higher than the rate of single Cantonese speaker.
#In all other province and territory, the rate of single Mandarin speaker is lower than the rate of single Cantonese speaker.

                          GEO  singlemandarinrate
0                Saskatchewan              0.1401
1                     Ontario              0.1231
2        Prince Edward Island              0.1121
3                      Canada              0.1119
4            British Columbia              0.1104
5                      Quebec              0.0936
6                    Manitoba              0.0925
7                     Alberta              0.0812
8               New Brunswick              0.0597
9       Northwest Territories              0.0588
10                Nova Scotia              0.0581
11  Newfoundland and Labrador              0.0429
                          GEO  singleCantoneserate
0                       Yukon               0.2281
1            British Columbia               0.1269
2                      Canada               0.1171
3                     Ontario               0.1154
4                    Manitoba               0.1099
5                      Quebec               

**Question 3 (4 marks)**

What kind of data would be interesting to receive in combination with the data in this table? Discuss how you could use this additional information to extend one of your guiding questions.

The Multiple responses of knowledge of languages is a bit vague. If it can provide the number of exact combination of languages spoken. It would increase the accuracy of analysis. For my first guiding question, it could be some statistic that speak both Mandarin and Cantonese that can be accuracy of iterference analysis. If the additional information is given, I can extend my guiding questions as how many Mandarin speaker also speaks Cantonese. It is a more interesting guiding question that I am interested in studying.

## Part D: Reflection (5 marks)

In 100 to 250 words, identify a concept you have found difficult or confusing from this assignment. Reflect on how your previous learning or experience helped you to understand this concept. Provide your reflection using markdown in the cell below

In this assignment, Join table is the concept that I find difficult. I had learned SQL in my undergrad days, but after that class, I never used SQL again, and in this assignment, I used SQL again after a few years, although I think I quickly recalled what I had learned in the past, such as basic syntax and basic SQL operations. But the complex concept of JOIN confused me. For example, in the last two questions of part B, I couldn't think of a solution that could be solved without using JOIN TABLE. So I found some SQL information on the Internet, which made me understand the concept of JOIN Table again. So after some practice, I solved the last two problems of part B by using JOIN method. And I used the JOIN method to solve my own problem in part C later on.

In [35]:
# Use this cell to include some code to dispose of your SQLAlchemy engine object
engine.dispose()

## References

Statistics Canada (Government of Canada). 2022. Knowledge of languages by age and gender: Canada, provinces and territories, census metropolitan areas and census agglomerations with parts. Retrieved October 15, 2022 from https://www150.statcan.gc.ca/t1/tbl1/en/tv.action?pid=9810021701